libraries/ESP8266WebServer/src/Parsing-impl.h

```diff
@@ -440,74 +441,61 @@ bool ESP8266WebServerTemplate<ServerType>::_parseForm(ClientType& client, const
-                    _currentUpload->type.c_str(),
-                    (int)_currentUpload->totalSize);
-                line = client.readStringUntil(0x0D);
-                client.readStringUntil(0x0A);
-                if (line == "--"){
-                  DBGWS("Done Parsing POST\n");
-                  break;
+                if ((strncmp((const char*)boundBuf, "--", 2) == 0) && (strcmp((const char*)(boundBuf + 2), boundary.c_str()) == 0))
+                    break;   //found the boundary, done parsing this file
+                if (first) first = false;   //only add newline characters after the first line
+                else {
+                    _uploadWriteByte('\r');
+                    _uploadWriteByte('\n');
                 }
-                continue;
-              } else {
-                _uploadWriteByte(0x0D);
-                _uploadWriteByte(0x0A);
-                _uploadWriteByte((uint8_t)('-'));
-                _uploadWriteByte((uint8_t)('-'));
-                uint32_t i = 0;
-                while(i < boundary.length()){
-                  _uploadWriteByte(endBuf[i++]);
+                // current line does not contain boundary, upload all bytes in boundary buffer
+                for (int j = 0; j < i; j++)
+                    _uploadWriteByte(boundBuf[j]);
+                // the initial pass (filling up the boundary buffer) did not reach the end of the line. Upload the rest of the line now
+                if (i >= 2 + bLen) {
+                    argByte = _uploadReadByte(client);
+                    while (argByte != '\r') {
+                        if (!client.connected()) return _parseFormUploadAborted();
+                        _uploadWriteByte(argByte);
+                        argByte = _uploadReadByte(client);
+                    }
                 }
-                argByte = _uploadReadByte(client);
-                goto readfile;
-              }
-            } else {
-              _uploadWriteByte(0x0D);
-              goto readfile;
+                _uploadReadByte(client); // '\n'
+            }
+            //Found the boundary string, finish processing this file upload
+            if (_currentHandler && _currentHandler->canUpload(_currentUri))
+                _currentHandler->upload(*this, _currentUri, *_currentUpload);
+            _currentUpload->totalSize += _currentUpload->currentSize;
+            _currentUpload->status = UPLOAD_FILE_END;
+            if (_currentHandler && _currentHandler->canUpload(_currentUri))
+                _currentHandler->upload(*this, _currentUri, *_currentUpload);
+            DBGWS("End File: %s Type: %s Size: %d\n",
+                _currentUpload->filename.c_str(),
+                _currentUpload->type.c_str(),
+                (int)_currentUpload->totalSize);
+            line = client.readStringUntil('\r');
+            client.readStringUntil('\n');
+            if (line == "--") {     // extra two dashes mean we reached the end of all form fields
```

Comment:
This now seems to be the only way to get out of this function.
Is it guaranteed there is this `--` line?
What if it got lost either due to lost connection, or buffers not long enough to receive all data?

Reply:
Yes, I believe '--' is the standard for ending multi-part form data so that line will always be there. There is a `if (!client.connected()) return _parseFormUploadAborted();` check before reading data from the client, so that's how the function will exit if the client disconnects. I just committed a few more of these lines that I missed the first time. I'm not sure what you mean about the buffers; large uploads will be processed in chunks of 2048 bytes. I have tested with files up to 100MB. They take a long time to upload but they don't cause any errors.

Reply:
Chunks of 2k will work OK I guess.
I've seen sometimes POST requests did not always have all data as it was somehow truncated or already re-used for something else.
But I guess for a file upload it may be dealt with differently.
It is just that I would also expect some timeout per transfer chunk, or else it may be stalled forever or WDT reboot if the connection is lost during transfer.

Reply:
I haven't looked in-depth at the rest of the code; I just used what was in place from the former implementation of form parsing. That being said, I don't think my code should cause any problems with timing out or missing data. We're only reading one byte at a time from the client, and each time we're checking if the client is still connected. If it disconnects then we abort the transfer. If the client data is being clobbered before we read it, that sounds like a bug somewhere else. This PR is just to fix the parsing bug. I haven't noticed any problems during my testing, but if you think there is a fundamental problem in how we handle POST data then we should work on that as a separate issue.